Modal 컴포넌트 구현

silvertae opened the pull request:
## Description
- 공통 컴포넌트 중 Modal 컴포넌트를 구현하였습니다.

## Key changes
![Modal2](https://github.com/masters2023-project-01-second-hand/second-hand-max-FE/assets/76121068/50b0faaf-0881-4bde-90cc-a49ed752feb4)

- components 폴더에 `Modal.tsx` 가 추가되었습니다. 그리고 테스트용으로 `TestModalContent`라는 임시 컴포넌트도 추가했습니다.
- 기본적인 레이아웃을 잡으려다 보니 `App.tsx` 구조가 약간 변경되었습니다.
- reset.css에 `border-box` 관련 속성이 추가되었습니다.
- favicon을 코끼리 아이콘으로 변경하고 title도 코끼리 마켓으로 수정하였습니다.
- 뒷배경이 밋밋해 보여서 제맘대로 코끼리를 추가해보았습니다.
- 버튼들 구현은 Jay가 만들어주신 공통 컴포넌트 Button을 활용했습니다.

## To reviewers
- Jay PR 머지된 dev랑 미리 머지시켜서 충돌되는 부분은 수정하였습니다. 이미 만들어진 Button 컴포넌트랑 색깔이름 수정도 vscode에서 미리 하면 좋을것 같아서요.
- 한 블로그글을 참고해서 dialog element를 활용해서 구현해보았고, 기본적으로 제공되는 `showModal`, `close` 메서드를 활용하다보니 useRef를 사용하게되었네요. (https://dev.to/link2twenty/react-using-native-dialogs-to-make-a-modal-popup-4b25)
- Modal의 구조는 `Header` 영역과 `Content` 영역으로 나누었는데, `Content`는 쓰이는 곳마다 다르다보니 `children`으로 받게 되었고, `Header` 부분은 제목(headline)이 있을 때(동네설정, 카테고리 모달)와 없을 때(동네 검색 모달) 이렇게 두 경우만 있는 것 같아서 headline props를 받아서 내부적으로 처리해보았습니다.
- 현재 구현된 모달은 추후에 모달안에 중첩해서 알럿창을 띄워줘야 하다보니 수정될 수 있습니다. 이 부분은 Nested Modal에 대해 좀 더 조사해보고 Alert 구현하면서 말씀드리겠습니다.

## Issue
- #7

silvertae committed:
design #7: reset.css 파일에 border-box 속성 추가

silvertae committed:
feat #7: Modal 컴포넌트 구현

- dialog element를 사용해서 구현
- 상위컴포넌트에서 props로 isOpen, onClose를 받아서 사용
- 열고 닫힐때 애니메이션 적용

silvertae committed:
chore #7: kiwi 아이콘 삭제 후 elephant 아이콘으로 favicon 변경

silvertae committed:
design #7: App.tsx에서 기본적인 레이아웃 구현후 Modal 테스트

- App에 AppContainer, Layout 이라는 styled components 추가
- 배경이 심심해서 코끼리 넣어보았음

silvertae committed:
feat #7: Modal 컴포넌트에서 headline props를 받아서 렌더링하는 기능 추가

- headline이 유무에 따라 padding이 변경됨
- 모달 내부 content는 children으로 받아서 렌더링

silvertae committed:
Merge branch 'dev' into feat/#7-Modal

silvertae committed:
refactor #7: 기존 버튼들 Button 컴포넌트로 리팩터링

- 공통 컴포넌트 Button 컴포넌트 적용
- theme에서 바뀐 color명으로 수정

altmit commented on src/App.tsx:
```diff
@@ -1,25 +1,72 @@
+}`;
+
+const Layout = styled.div`
+  width: 393px;
+  height: 852px;
```
우선 피그마에 적용되어 있는 사이즈를 적용한 것이라 생각이 듭니다.
저희 프로젝트가 `모바일 웹` 이기 때문에 약간의 수정이 필요하다고 판단됩니다.

제 생각은 접속한 유저의 디바이스를 체크하거나 유저의 화면 크기로 판단하여 모바일 or 웹을 판단하고
만약 웹이라면 지금 처럼 저희가 정한 임의의 값을 px로 작성하고 모바일 이라면 width, height 100vw로 하면 좋을 것 같네요.

이 부분은 추후에 메인 작업을 시작할 때 같이 고민하면 좋을 것 같습니다.

silvertae replied:
모바일 디바이스에서 접속하는 경우는 생각하지 못했는데 말씀해주신 방법이 좋을 것 같네요

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+  useEffect,
+  useRef,
+} from 'react';
+import { styled, keyframes } from 'styled-components';
```
styled components에 keyframes가 있군요 처음 알았네요

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+  const modalRef = useRef<HTMLDialogElement>(null);
+  const hasHeadline = !!headline;
+  // Esc로 닫았을 때는 close 이벤트 감지해서 애니메이션 적용
+  const onCancel: MouseEventHandler<HTMLDialogElement> = useCallback(
```
대부분에 함수에 useCallback을 사용하셨는데 사용하신 이유가 있나요?

저는 open이라는 값으로 랜더링이 충분히 컨트롤 되고 있다고 판단되어서 useCallback을 쓰지 않았다고 불필요한 랜더링이 안 생길 것 같네요

silvertae replied:
참고한 코드에서 useCallback을 사용했길래 그냥 유지했는데, 제가 useCallback을 이번에 처음 사용해봐서 렌더링 할 때 성능상 큰 이점이 있는지는 좀 더 찾아보고 말씀드릴 수 있을 것 같네요.

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+  // 모달 바깥 영역 클릭시 닫히도록
+  const onClick: MouseEventHandler<HTMLDialogElement> = useCallback(
+    ({ target }) => {
+      const { current: el } = modalRef;
```
current를 el로 구조분해한 이유가 있을까요?
element라고 인지 시켜주고 싶은 이유였다면 el보단 풀네임이 좋을 것 같네요.

그리고
```ts
const dialog= modalRef.current
``` 
저는 이렇게 많이 사용해서 인지  조금 낯설게 느껴지네요.

silvertae replied:
다시보니 굳이 구조분해 할 필요가 없을 것 같네요. 말씀해주신 것처럼 ref.current에를 변수로 선언해서 사용하는게 저도 더 좋아보입니다.
(그리고 el은 풀네임 위주로 쓰기로 했었는데 줄여쓰는게 익숙하다보니 까먹었네요 😅)

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+    if (el && open) {
+      el.showModal();
+    }
+  }, [open]);
```
```ts
useEffect(() => {
  const dialog = modalRef.current;
  
  open && dialog?.showModal();
}, [open]);
```

이렇게도 작성 가능할 것 같네요.

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+
+  return (
+    <Dialog
+      className={open ? '' : 'modal--closing'}
```
저는 react + styled component의 장점 중 하나가 id, className을 할당하지 않아도 되는 것이라고 생각해요.

물론 쓴다고 나쁜것 이라고 생각은 안하는데 없어도 가능한 로직이라면 없는게 좋을 것 같네요.

open이 false일 경우 Modal 컴포넌트를 랜더링 하지 않으면 없어도 되는 className과 css 애니메이션이지 않을까? 생각 합니다.

만약 그렇게 한다면 닫기 버튼, esc 등 모달이 닫혀야 하는 로직에서 애니메이션이 끝난 뒤 onClose 함수를 실행해 open을 false로 변경되게 해야 할 것 같네요

silvertae replied:
저도 className 사용을 웬만해선 지양하려는 편인데 모달이 닫힐때 애니메이션 구현을 위해서는 어쩔 수 없이 필요할 것 같더라구요.

이전에 비슷한 상황이 있었는데 보통 컴포넌트가 unmount 되는 경우 애니메이션 적용이 잘 안 됐던 경험이 있었습니다.

제이가 말씀해주신 방법으로 한번 다시 시도해보고 말씀드리겠습니다.

altmit commented on src/App.tsx:
```diff
@@ -1,25 +1,72 @@
+        >
+          Open Modal
+        </Button>
+        <Modal
+          open={isOpen}
+          onClose={() => setIsOpen(false)}
+          headline="동네 설정"
+        >
+          <TestContent />
+        </Modal>
```
Modal 컴포넌트 자체를 isOpen일 경우만 랜더링 하는 건 어떨까요?

```ts
{ isOpen && 
    <Modal
      open={isOpen}
      onClose={() => setIsOpen(false)}
      headline="동네 설정"
    >
      <TestContent />
    </Modal>
}
```

silvertae replied:
위에 코멘트와 연결될 수도 있는 부분인것 같아 같이 적용해보겠습니다.

altmit commented on src/components/Modal.tsx:
```diff
@@ -0,0 +1,174 @@
+import { Button } from './Button';
+
+type ModalProps = DialogHTMLAttributes<HTMLDialogElement> & {
+  open: boolean;
```
밖에서는 isOpen 이라는 이름으로 관리 하는데 여기는 그냥 open인 이유가 있나요?

silvertae replied:
다른 이름일 필요는 없을것 같네요 수정하겠습니다

altmit reviewed (approved):
Modal 컴포넌트 구현하시느라 수고 하셨습니다.
제 생각을 코멘트로 조금 남겨 봤습니다.

단순 저의 생각일 뿐 정답이나 꼭 그렇게 했으면 하는 코멘트는 아닙니다.
퓨즈의 의견도 듣고 정하면 될 것 같습니다.

그리고 드릴 말씀이 있는데 이번 pr과는 상관 없지만 
 import 순서 정렬해 주는 부분에서 eslint가 경고를 보여주고 있습니다.
 
![1](https://github.com/masters2023-project-01-second-hand/second-hand-max-FE/assets/41321198/53de6e0a-6b14-4be6-a7e0-5041e5db2558)

그래서 왜 이런 경고를 하는지 gpt에게 물어 봤습니다.

> 코드 스타일 가이드에서는 종종 내부 패키지 또는 현재 디렉토리에 가까운 모듈을 먼저 임포트하고, 그 다음 외부 패키지 또는 프로젝트의 더 먼 경로에 있는 모듈을 임포트하는 것을 권장합니다. 이는 코드의 가독성과 유지보수를 쉽게 하기 위한 것입니다.

silvertae committed:
refactor #7: 리뷰어 수정사항 반영

- isOpen props 이름 통일
- modalRef에서 불필요한 destructuring 제거
- Modal 컴포넌트 주석 제거

silvertae committed:
fix #7: eslint rule에서 import문 순서 수정

silvertae committed:
style #7: export default를 export로 변경

silvertae committed:
feat #7: Modal 컴포넌트에 createPortal 적용

- Modal 컴포넌트에 createPortal 반환값을 리턴하도록 수정
- Modal이 들어갈 위치로 index.html에 div#modal-root를 추가

silvertae committed:
refactor #7: Modal이 isOpen 상태에 따라 조건부 렌더링 되도록 수정

silvertae commented:
@altmit 
말씀해주신 피드백 반영하였고 자잘한 수정들도 포함되었습니다.
그리고 변경점은 createPortal로 index.html에서 div#modal-root에서 렌더링 되도록 하였고, isOpen에 상태에 따라서 조건부 렌더링 되도록 했습니다. 따라서 현재 unmount시 애니메이션이 적용되지 않는데 이 문제는 다른 이슈를 파서 추후에 따로 작업해볼 예정입니다.

silvertae committed:
Merge branch 'dev' into feat/#7-Modal

silvertae committed:
style #7: 단순한 if문은 중괄호를 생략하도록 수정
